File: Engr-T.stark/wordBreak.cpp
class Solution {
 public:

set<string> set;

bool func(string s, int idx, vector<int>& dp)
{
    if(idx==s.size())
        return true;
    if(dp[idx]!=-1)
        return dp[idx];
    
    string str;
    
    for(int i=idx;i<s.size();i++)
    {
        str += s[i];
        if(set.find(str)!=set.end())
        {
            if(func(s,i+1,dp))
                return true;
        }
    }
    return dp[idx] = false;
}

bool wordBreak(string s, vector<string>& wordDict) {
    
    vector<int> dp(s.size(),-1);
    for(auto str: wordDict)
        set.insert(str);
    
    return func(s,0,dp);   
}
};

File: Eff_Emm/word_break.py
class Solution(object):
    def wordBreak(self, s, wordDict):
        """
        :type s: str
        :type wordDict: List[str]
        :rtype: bool
        """
        
        
        dp = [False] * (len(s) +1)
        dp[len(s)] = True
        
        for i in range(len(s)-1,-1,-1):
            for w in  wordDict:
                if (i+ len(w)) <= len(s) and s[i: i+ len(w)] == w:
                    dp[i] = dp[i + len(w)]
                if dp[i]:
                    break
        return dp[0]
               
         